Source/FieldSolver/FiniteDifferenceSolver/HybridPICSolveE.cpp

```diff
@@ -0,0 +1,511 @@
+        lev, hybrid_model, a_dt_type
+    );
+    amrex::Abort(Utils::TextMsg::Err(
+        "currently hybrid E-solve does not work for RZ"));
+}
+
+#else
+
+template<typename T_Algo>
+void FiniteDifferenceSolver::HybridPICSolveECartesian (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Efield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Jfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Jifield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield,
+    std::unique_ptr<amrex::MultiFab> const& rhofield,
+    std::unique_ptr<amrex::MultiFab> const& Pefield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& edge_lengths,
+    int lev, std::unique_ptr<HybridPICModel> const& hybrid_model,
+    DtType a_dt_type )
+{
+#ifndef AMREX_USE_EB
+    amrex::ignore_unused(edge_lengths);
+#endif
+
+    // for the profiler
+    amrex::LayoutData<amrex::Real>* cost = WarpX::getCosts(lev);
+
+    using namespace ablastr::coarsen::sample;
+
+    // get hybrid model parameters
+    const auto eta = hybrid_model->m_eta;
+    const auto rho_floor = hybrid_model->m_n_floor * PhysConst::q_e;
+
+    // Index type required for interpolating fields from their respective
+    // staggering to the Ex, Ey, Ez locations
+    amrex::GpuArray<int, 3> const& Ex_stag = hybrid_model->Ex_IndexType;
+    amrex::GpuArray<int, 3> const& Ey_stag = hybrid_model->Ey_IndexType;
+    amrex::GpuArray<int, 3> const& Ez_stag = hybrid_model->Ez_IndexType;
+    amrex::GpuArray<int, 3> const& Jx_stag = hybrid_model->Jx_IndexType;
+    amrex::GpuArray<int, 3> const& Jy_stag = hybrid_model->Jy_IndexType;
+    amrex::GpuArray<int, 3> const& Jz_stag = hybrid_model->Jz_IndexType;
+    amrex::GpuArray<int, 3> const& Bx_stag = hybrid_model->Bx_IndexType;
+    amrex::GpuArray<int, 3> const& By_stag = hybrid_model->By_IndexType;
+    amrex::GpuArray<int, 3> const& Bz_stag = hybrid_model->Bz_IndexType;
+
+    // Parameters for `interp` that maps from Yee to nodal mesh and back
+    amrex::GpuArray<int, 3> const& nodal = {1, 1, 1};
+    // The "coarsening is just 1 i.e. no coarsening"
+    amrex::GpuArray<int, 3> const& coarsen = {1, 1, 1};
+
+    // The E-field calculation is done in 2 steps:
+    // 1) The J x B term is calculated on a nodal mesh in order to ensure
+    //    energy conservation.
+    // 2) The nodal E-field values are averaged onto the Yee grid and the
+    //    electron pressure & resistivity terms are added (these terms are
+    //    naturally located on the Yee grid).
+
+    // Create a temporary multifab to hold the nodal E-field values
+    // Note the multifab has 3 values for Ex, Ey and Ez which we can do here
+    // since all three components will be calculated on the same grid.
+    auto const& ba = convert(rhofield->boxArray(), IntVect::TheNodeVector());
+    MultiFab enE_nodal_mf(ba, rhofield->DistributionMap(), 3, rhofield->nGrow());
```

Comment:
Naively, it seemed to me that `enE_nodal_mf` actually does not need to have guard cells.
```suggestion
    MultiFab enE_nodal_mf(ba, rhofield->DistributionMap(), 3, IntVect::TheZeroVector);
```
since we then only interpolate it to the staggered positions of E than are *inside* the area delimited by the outermost nodes of the valid grid. Is that correct?
In this case, we could get rid of the ` enE_nodal_mf.FillBoundary` further below.

Reply:
That reasoning makes sense to me, thanks for pointing it out! I've made this change, presumably the CI tests will all still pass.

Reply:
This improvement cut the field solve time in half for large GPU runs :tada: